File: front50-gcs/src/main/java/com/netflix/spinnaker/front50/model/GcsStorageService.java
/*
 * Copyright 2016 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.spinnaker.front50.model;

import com.google.common.annotations.VisibleForTesting;
import com.netflix.spinnaker.front50.exception.NotFoundException;

import com.google.api.services.storage.Storage;
import com.google.api.services.storage.StorageScopes;
import com.google.api.services.storage.model.*;

import com.google.api.services.storage.model.Bucket;
import com.google.api.client.http.ByteArrayContent;
import com.google.api.client.util.DateTime;
import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;

import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpResponseException;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import org.apache.commons.lang.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class GcsStorageService implements StorageService {
  private static final String DEFAULT_DATA_FILENAME = "specification.json";
  private static final String LAST_MODIFIED_FILENAME = "last-modified";
  private final Logger log = LoggerFactory.getLogger(getClass());

  private ObjectMapper objectMapper = new ObjectMapper();
  private String projectName;
  private String bucketName;
  private String basePath;
  private Storage storage;
  private Storage.Objects obj_api;
  private String dataFilename = DEFAULT_DATA_FILENAME;

  /**
   * Bucket location for when a missing bucket is created. Has no effect if the bucket already
   * exists.
   */
  private String bucketLocation;

  public Storage getStorage() { return this.storage; }
  public ObjectMapper getObjectMapper() { return this.objectMapper; }

  private GoogleCredential loadCredential(
      HttpTransport transport, JsonFactory factory, String jsonPath)
      throws IOException {
    GoogleCredential credential;
    if (!jsonPath.isEmpty()) {
      FileInputStream stream = new FileInputStream(jsonPath);
      credential = GoogleCredential.fromStream(stream, transport, factory)
                      .createScoped(Collections.singleton(StorageScopes.DEVSTORAGE_FULL_CONTROL));
      log.info("Loaded credentials from from " + jsonPath);
    } else {
      log.info("spinnaker.gcs.enabled without spinnaker.gcs.jsonPath. " +
                   "Using default application credentials. Using default credentials.");
      credential = GoogleCredential.getApplicationDefault();
    }
    return credential;
  }

  @VisibleForTesting
  GcsStorageService(String bucketName,
                    String bucketLocation,
                    String basePath,
                    String projectName,
                    Storage storage) {
    this.bucketName = bucketName;
    this.bucketLocation = bucketLocation;
    this.basePath = basePath;
    this.projectName = projectName;
    this.storage = storage;
    this.obj_api = storage.objects();
  }

  public GcsStorageService(String bucketName,
                           String bucketLocation,
                           String basePath,
                           String projectName,
                           String credentialsPath,
                           String applicationVersion) {
    this(bucketName,
         bucketLocation,
         basePath,
         projectName,
         credentialsPath,
         applicationVersion,
         DEFAULT_DATA_FILENAME);
  }

  public GcsStorageService(String bucketName,
                           String bucketLocation,
                           String basePath,
                           String projectName,
                           String credentialsPath,
                           String applicationVersion,
                           String dataFilename) {
    Storage storage;

    try {
      HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
      JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
      GoogleCredential credential = loadCredential(httpTransport, jsonFactory,
                                                   credentialsPath);

      String applicationName = "Spinnaker-front50/" + applicationVersion;
      storage = new Storage.Builder(httpTransport, jsonFactory, credential)
                           .setApplicationName(applicationName)
                           .build();
    } catch (IOException|java.security.GeneralSecurityException e) {
        throw new IllegalStateException(e);
    }

    this.bucketName = bucketName;
    this.bucketLocation = bucketLocation;
    this.basePath = basePath;
    this.projectName = projectName;
    this.storage = storage;
    this.obj_api = this.storage.objects();
    this.dataFilename = dataFilename;
  }

  /**
   * Check to see if the bucket exists, creating it if it is not there.
   */
  public void ensureBucketExists() {
    try {
      Bucket bucket = storage.buckets().get(bucketName).execute();
    } catch (HttpResponseException e) {
      if (e.getStatusCode() == 404) {
        log.warn("Bucket {} does not exist. Creating it in project={}",
                 bucketName, projectName);
        Bucket.Versioning versioning = new Bucket.Versioning().setEnabled(true);
        Bucket bucket = new Bucket().setName(bucketName).setVersioning(versioning);
        if (StringUtils.isNotBlank(bucketLocation)) {
          bucket.setLocation(bucketLocation);
        }
        try {
            storage.buckets().insert(projectName, bucket).execute();
        } catch (IOException e2) {
            log.error("Could not create bucket={} in project={}: {}",
                      bucketName, projectName, e2);
            throw new IllegalStateException(e2);
        }
      } else {
          log.error("Could not get bucket={}: {}", bucketName, e);
          throw new IllegalStateException(e);
      }
    } catch (IOException e) {
        log.error("Could not get bucket={}: {}", bucketName, e);
        throw new IllegalStateException(e);
    }
  }

  /**
   * Returns true if the storage service supports versioning.
   */
  public boolean supportsVersioning() {
    try {
      Bucket bucket = storage.buckets().get(bucketName).execute();
      Bucket.Versioning v = bucket.getVersioning();
      return v != null && v.getEnabled();
    } catch (IOException e) {
      throw new IllegalStateException(e);
    }
  }

  public <T extends Timestamped> T
         loadCurrentObject(String objectKey, String daoTypeName, Class<T> clas)
         throws NotFoundException {
    String path = keyToPath(objectKey, daoTypeName);
    try {
      StorageObject storageObject = obj_api.get(bucketName, path).execute();
      T item = deserialize(storageObject, clas, true);
      item.setLastModified(storageObject.getUpdated().getValue());
      log.debug("Loaded bucket={} path={}", bucketName, path);
      return item;
    } catch (HttpResponseException e) {
      log.error("Failed to load {} {}: {} {}",
                daoTypeName, objectKey, e.getStatusCode(), e.getStatusMessage());
      if (e.getStatusCode() == 404) {
          throw new NotFoundException(String.format("No file at path=%s", path));
      }
      throw new IllegalStateException(e);
    } catch (IOException e) {
      throw new IllegalStateException(e);

    }
  }

  public <T extends Timestamped> T
         loadObjectVersion(String objectKey, String daoTypeName, Class<T> clas,
                           String versionId) throws NotFoundException {
      return null;
  }

  public void deleteObject(String objectKey, String daoTypeName) {
    String path = keyToPath(objectKey, daoTypeName);
    try {
      obj_api.delete(bucketName, path).execute();
      writeLastModified(daoTypeName);
    } catch (IOException e) {
      log.error("Delete failed on path={}", path);
      throw new IllegalStateException(e);
    }
  }

  public <T extends Timestamped>
         void storeObject(String objectKey, String daoTypeName, T obj) {
    String path = keyToPath(objectKey, daoTypeName);
    try {
      byte[] bytes = objectMapper.writeValueAsBytes(obj);
      StorageObject object = new StorageObject().setBucket(bucketName).setName(path);
      ByteArrayContent content = new ByteArrayContent("application/json", bytes);
      obj_api.insert(bucketName, object, content).execute();
      writeLastModified(daoTypeName);
    } catch (IOException e) {
      log.error("Update failed on path={}: {}", path, e);
      throw new IllegalStateException(e);
    }
  }

  public Map<String, Long> listObjectKeys(String daoTypeName) {
    String rootFolder = daoRoot(daoTypeName);
    int skipToOffset = rootFolder.length() + 1;  // + Trailing slash
    int skipFromEnd = dataFilename.length() + 1;  // + Leading slash

    Map<String, Long> result = new HashMap<String, Long>();
    log.debug("Listing {}", daoTypeName);
    try {
      Storage.Objects.List listObjects = obj_api.list(bucketName);
      listObjects.setPrefix(rootFolder);
      com.google.api.services.storage.model.Objects objects;
      do {
          objects = listObjects.execute();
          List<StorageObject> items = objects.getItems();
          if (items != null) {
              for (StorageObject item: items) {
                  String name = item.getName();
                  if (name.endsWith(dataFilename)) {
                      result.put(name.substring(skipToOffset,
                                                name.length() - skipFromEnd),
                                 item.getUpdated().getValue());
                  }
              }
          }
          listObjects.setPageToken(objects.getNextPageToken());
      } while (objects.getNextPageToken() != null);
    } catch (IOException e) {
      log.error("Could not fetch items from Google Cloud Storage: {}", e);
      return new HashMap<String, Long>();
    }

    return result;
  }

  public <T extends Timestamped> Collection<T>
         listObjectVersions(String objectKey, String daoTypeName, Class<T> clas,
                            int maxResults) throws NotFoundException {
    String path = keyToPath(objectKey, daoTypeName);
    ArrayList<T> result = new ArrayList<T>();
    try {
      Storage.Objects.List listObjects = obj_api.list(bucketName)
          .setPrefix(path)
          .setVersions(true)
          .setMaxResults(new Long(maxResults));
      com.google.api.services.storage.model.Objects objects;
      do {
          objects = listObjects.execute();
          List<StorageObject> items = objects.getItems();
          if (items != null) {
            for (StorageObject item : items) {
                T have = deserialize(item, clas, false);
                if (have != null) {
                  result.add(have);
                }
            }
          }
          listObjects.setPageToken(objects.getNextPageToken());
      } while (objects.getNextPageToken() != null);
    } catch (IOException e) {
      log.error("Could not fetch versions from Google Cloud Storage: {}", e);
      return new ArrayList<>();
    }
    return result;
  }

  private <T extends Timestamped> T
          deserialize(StorageObject object, Class<T> clas, boolean current_version)
      throws java.io.UnsupportedEncodingException {
    try {
        ByteArrayOutputStream output = new java.io.ByteArrayOutputStream();
        Storage.Objects.Get getter = obj_api.get(object.getBucket(), object.getName());
        if (!current_version) {
            getter.setGeneration(object.getGeneration());
        }
        getter.executeMediaAndDownloadTo(output);
        String json = output.toString("UTF8");
        return objectMapper.readValue(json, clas);
      } catch (IOException ex) {
        if (current_version) {
          log.error("Error reading {}: {}", object.getName(), ex);
        } else {
          log.error("Error reading {} generation={}: {}",
                    object.getName(), object.getGeneration(), ex);
        }
        return null;
    }
  }

  private void writeLastModified(String daoTypeName) {
      // We'll just touch the file since the StorageObject manages a timestamp.
      String timestamp_path = daoRoot(daoTypeName) + '/' + LAST_MODIFIED_FILENAME;
      StorageObject object = new StorageObject()
          .setBucket(bucketName)
          .setName(timestamp_path)
          .setUpdated(new DateTime(System.currentTimeMillis()));
      try {
        obj_api.update(bucketName, object.getName(), object).execute();
      } catch (HttpResponseException e) {
        log.error("writeLastModified failed to update: {}", e.toString());

        if (e.getStatusCode() == 404 || e.getStatusCode() == 400) {
            byte[] bytes = "{}".getBytes();
            ByteArrayContent content = new ByteArrayContent("application/json", bytes);

            try {
              obj_api.insert(bucketName, object, content).execute();
            } catch (IOException ioex) {
              log.error("writeLastModified insert failed too: {}", ioex);
              throw new IllegalStateException(e);
            }
        } else {
            throw new IllegalStateException(e);
        }
      } catch (IOException e) {
        log.error("writeLastModified failed: {}", e);
        throw new IllegalStateException(e);
      }

      try {
          // If the bucket is versioned, purge the old timestamp versions
          // because they serve no value and just consume storage and extra time
          // if we eventually destroy this bucket.
          purgeOldVersions(timestamp_path);
      } catch (IOException e) {
          log.error("Failed to purge old versions of {}. Ignoring error.",
                    timestamp_path);
      }
  }

  // Remove the old versions of a path.
  // Versioning is per-bucket but it doesnt make sense to version the
  // timestamp objects so we'll aggressively delete those.
  private void purgeOldVersions(String path) throws IOException {
      Storage.Objects.List listObjects = obj_api.list(bucketName)
          .setPrefix(path)
          .setVersions(true);

      com.google.api.services.storage.model.Objects objects;

      // Keep the 0th object on the first page (which is current).
      List<Long> generations = new ArrayList(32);
      do {
          objects = listObjects.execute();
          List<StorageObject> items = objects.getItems();
          if (items != null) {
              int n = items.size();
              while (--n >= 0) {
                  generations.add(items.get(n).getGeneration());
              }
          }
          listObjects.setPageToken(objects.getNextPageToken());
      } while (objects.getNextPageToken() != null);

      for (long generation : generations) {
          if (generation == generations.get(0)) {
              continue;
          }
          log.debug("Remove {} generation {}", path, generation);
          obj_api.delete(bucketName, path).setGeneration(generation).execute();
      }
  }

  public long getLastModified(String daoTypeName) {
      String path = daoRoot(daoTypeName) + '/' + LAST_MODIFIED_FILENAME;
      try {
          return obj_api.get(bucketName, path).execute().getUpdated().getValue();
      } catch (HttpResponseException e) {
          if (e.getStatusCode() == 404) {
              log.error("No timestamp file at {}. Creating a new one.", path);
              writeLastModified(daoTypeName);
              return 0L;
          }
          log.error("Error writing timestamp file {}", e.toString());
          return 0L;
      } catch (Exception e) {
          log.error("Error accessing timestamp file {}", e.toString());
          return 0L;
      }
  }

  private String daoRoot(String daoTypeName) {
      return basePath + '/' + daoTypeName;
  }

  private String keyToPath(String key, String daoTypeName) {
      return daoRoot(daoTypeName) + '/' + key + '/' + dataFilename;
  }
}

File: front50-gcs/src/main/kotlin/com/netflix/spinnaker/front50/model/GcsStorageService.kt
/*
 * Copyright 2020 Google, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.netflix.spinnaker.front50.model

import com.fasterxml.jackson.databind.ObjectMapper
import com.google.cloud.storage.Blob
import com.google.cloud.storage.BlobId
import com.google.cloud.storage.BlobInfo
import com.google.cloud.storage.BucketInfo
import com.google.cloud.storage.Storage
import com.google.cloud.storage.Storage.BlobListOption
import com.google.cloud.storage.Storage.BucketField
import com.google.cloud.storage.Storage.BucketGetOption
import com.google.cloud.storage.StorageException
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.netflix.spinnaker.front50.exception.NotFoundException
import java.io.IOException
import java.util.Comparator
import java.util.concurrent.Executor
import java.util.stream.Collectors
import org.slf4j.LoggerFactory

class GcsStorageService(
  private val storage: Storage,
  private val bucketName: String,
  private val bucketLocation: String,
  private val basePath: String,
  private val dataFilename: String,
  private val objectMapper: ObjectMapper,
  private val executor: Executor
) : StorageService {

  companion object {
    private val log = LoggerFactory.getLogger(GcsStorageService::class.java)
    private const val LAST_MODIFIED_FILENAME = "last-modified"
  }

  private val modTimeState = ObjectType.values().map { it to ModificationTimeState(it) }.toMap()

  fun ensureBucketExists() {
    val bucket = storage.get(bucketName)
    if (bucket == null) {
      val bucketInfo = BucketInfo.newBuilder(bucketName)
        .setVersioningEnabled(true)
      if (bucketLocation.isNotBlank()) {
        bucketInfo.setLocation(bucketLocation)
      }
      storage.create(bucketInfo.build())
    }
  }

  override fun supportsVersioning(): Boolean {
    val bucket = storage.get(bucketName, BucketGetOption.fields(BucketField.VERSIONING))
    return bucket.versioningEnabled() == true
  }

  override fun <T : Timestamped> loadObject(objectType: ObjectType, objectKey: String): T {
    try {
      val blobId = blobIdForKey(objectType, objectKey)
      val blob = storage.get(blobId)
        ?: throw NotFoundException("Couldn't retrieve $objectType $objectKey from GCS")
      val bytes = storage.readAllBytes(blob.blobId)
      val obj: T = parseObject(bytes, objectType, objectKey)
      obj.lastModified = blob.updateTime
      return obj
    } catch (e: StorageException) {
      throw GcsStorageServiceException("error loading $objectType $objectKey", e)
    }
  }

  override fun deleteObject(objectType: ObjectType, objectKey: String) {
    try {
      if (storage.delete(blobIdForKey(objectType, objectKey))) {
        writeLastModified(objectType)
      }
    } catch (e: Exception) {
      throw GcsStorageServiceException("error deleting $objectType $objectKey", e)
    }
  }

  override fun <T : Timestamped?> storeObject(objectType: ObjectType, objectKey: String, item: T) {
    val blobId = blobIdForKey(objectType, objectKey)
    try {
      val bytes = objectMapper.writeValueAsBytes(item)
      storage.create(BlobInfo.newBuilder(blobId).setContentType("application/json").build(), bytes)
      writeLastModified(objectType)
    } catch (e: Exception) {
      throw GcsStorageServiceException("Error writing $objectType $objectKey", e)
    }
  }

  override fun listObjectKeys(objectType: ObjectType): MutableMap<String, Long> {
    val results = ImmutableMap.builder<String, Long>()

    try {
      val rootDirectory = daoRoot(objectType)
      storage.list(bucketName, BlobListOption.prefix("$rootDirectory/"))
        .iterateAll()
        .forEach { blob ->
          val objectKey = getObjectKey(blob, rootDirectory)
          if (objectKey != null) {
            results.put(objectKey, blob.updateTime)
          }
        }
    } catch (e: Exception) {
      throw GcsStorageServiceException("error listing $objectType objects", e)
    }

    return results.build()
  }

  private fun getObjectKey(blob: Blob, rootDirectory: String): String? {
    val name = blob.name
    return if (!name.startsWith("$rootDirectory/") || !name.endsWith("/$dataFilename")) {
      null
    } else name.substring(rootDirectory.length + 1, name.length - dataFilename.length - 1)
  }

  override fun <T : Timestamped> listObjectVersions(
    objectType: ObjectType,
    objectKey: String,
    maxResults: Int
  ): MutableCollection<T> {

    try {
      val path = pathForKey(objectType, objectKey)
      val listResults = storage.list(bucketName, BlobListOption.prefix(path), BlobListOption.versions(true))
      val blobs: List<Blob> = ImmutableList.copyOf(listResults.iterateAll())
      return blobs.stream()
        .filter { blob: Blob -> blob.name == path }
        .sorted(Comparator.comparing { blob: Blob -> blob.updateTime }.reversed())
        .limit(maxResults.toLong())
        .map { blob: Blob ->
          parseObject<T>(blob.getContent(), objectType, objectKey)
            .apply { lastModified = blob.updateTime }
        }
        .collect(Collectors.toList())
    } catch (e: Exception) {
      throw GcsStorageServiceException("error loading $objectType $objectKey", e)
    }
  }

  override fun getLastModified(objectType: ObjectType): Long {
    try {
      val blob = storage.get(lastModifiedBlobId(objectType))
      if (blob == null) {
        writeLastModified(objectType)
        return 0
      } else {
        return blob.updateTime
      }
    } catch (e: Exception) {
      throw GcsStorageServiceException("error loading timestamp for $objectType objects", e)
    }
  }

  private fun writeLastModified(objectType: ObjectType) {
    val updaterState = modTimeState.getValue(objectType)
    updaterState.updateNeeded()
  }

  private fun updateLastModified(objectType: ObjectType, state: ModificationTimeState) {
    state.updateTaskStarted()
    val blobInfo = BlobInfo.newBuilder(lastModifiedBlobId(objectType)).build()
    try {
      // Calling update() is enough to change the modification time on the file, which is all we
      // care about. It doesn't matter if we don't actually specify any fields to change.
      storage.update(blobInfo)
    } catch (e: Exception) {
      when {
        e is StorageException && e.code == 404 ->
          try {
            storage.create(blobInfo)
          } catch (e: Exception) {
            log.warn("Error updating last modified time for $objectType", e)
          }
        else -> log.warn("Error updating last modified time for $objectType", e)
      }
    } finally {
      state.updateTaskFinished()
    }
  }

  private fun scheduleUpdateLastModified(objectType: ObjectType, updaterState: ModificationTimeState) {
    executor.execute { updateLastModified(objectType, updaterState) }
  }

  private inner class ModificationTimeState(private var objectType: ObjectType) {

    private var needsUpdate = false
    private var updateRunning = false

    /**
     * Specify that an object has been modified so the modification time for this objectType needs
     * to be updated.
     */
    @Synchronized
    fun updateNeeded() {
      val neededUpdate = needsUpdate
      needsUpdate = true
      // We need to check `neededUpdate` or else several near-simultaneous calls would schedule
      // multiple executions of the task before any of them have a chance to get up and running
      if (!neededUpdate && !updateRunning) {
        scheduleUpdateLastModified(objectType, this)
      }
    }

    /** Should be called by the update task to indicate it has started. */
    @Synchronized
    fun updateTaskStarted() {
      updateRunning = true
      needsUpdate = false
    }

    /** Should be called by the update task to indicate it has completed. */
    @Synchronized
    fun updateTaskFinished() {
      updateRunning = false
      // If another update request came in while the task was running, we need to launch the task
      // again.
      if (needsUpdate) {
        scheduleUpdateLastModified(objectType, this)
      }
    }
  }

  private fun <T : Timestamped> parseObject(bytes: ByteArray, objectType: ObjectType, objectKey: String): T {
    return try {
      @Suppress("UNCHECKED_CAST")
      objectMapper.readValue(bytes, objectType.clazz as Class<T>)
    } catch (e: IOException) {
      throw GcsStorageServiceException("error reading $objectType $objectKey", e)
    }
  }

  private fun blobIdForKey(objectType: ObjectType, key: String): BlobId {
    return BlobId.of(bucketName, pathForKey(objectType, key))
  }

  private fun pathForKey(objectType: ObjectType, key: String): String {
    return "${daoRoot(objectType)}/$key/$dataFilename"
  }

  private fun lastModifiedBlobId(objectType: ObjectType): BlobId {
    return BlobId.of(bucketName, "${daoRoot(objectType)}/$LAST_MODIFIED_FILENAME")
  }

  private fun daoRoot(objectType: ObjectType): String {
    return "$basePath/${objectType.group}"
  }
}

private class GcsStorageServiceException(message: String, cause: Throwable) : RuntimeException(message, cause)
